### ISSUE TITLE
cross compile support for aarch64?

### ISSUE THREAD

**Issue by younggam94** [ApolloAuto/apollo#12810]
We appreciate you go through Apollo documentations and search previous issues before creating an new one. If neither of the sources helped you with your issues, please report the issue using the following form. Please note missing info can delay the response time.

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 18.04)**: 18.04
- **Apollo installed from (source or binary)**: source
- **Apollo version (3.5, 5.0, 5.5, 6.0)**: 6.0
- **Output of `apollo.sh config` if on `master` branch**: 

### Steps to reproduce the issue:
- **Please use bullet points and include as much details as possible**:
1. run qemu container for aarch64 on top of x86(HOST) machine
2. run dev_start.sh with TARGET machine for aarch64.
3. run apollo.sh build_cpu

After step 3 and run cyber monitor, couldn't see any channel info in the cyber monitor.

However, we run step 2,3 on the aarch64 host machine, could see cyber monitor channel.

We don't find any clue yet. why cross compilation doesn't work.

Can somebody know the reason? or how can we do cross compilation for aarch64?

**Comment by storypku** [comment]
There is a `-m <arch>` option for `cyber_start.sh` in Apollo 6.0 and master, which support Running CyberRT container for aarch64 on x86_64 host. After `cyber_into.sh`, run `bash apollo.sh build cyber` to build cyber only. And then do your experiment there.

Please note that this feature is experimental, and may be removed in the future. 

As to cross-architecture support for `dev_start.sh`, the answer is: no, we won't support running Apollo aarch64 on x86_64 hosts.

**Comment by younggam94** [comment]
Thx for your info.
I did as you mentioned for cyber.
And checked example talker and listener.
Listener doesn't seem to work properly.

Here is the log talker and listener
talker log
Log file created at: 2020/10/20 10:10:06
Running on machine: in-cyber-docker
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I1020 10:10:06.432982 11584 global_data.cc:153] host ip: 192.168.10.31
W1020 10:10:06.488025 11584 scheduler_factory.cc:63] No sched conf found, use default conf.
I1020 10:10:06.496665 11594 processor.cc:42] processor_tid: 11594
I1020 10:10:06.499819 11595 processor.cc:42] processor_tid: 11595
I1020 10:10:06.501027 11596 processor.cc:42] processor_tid: 11596
I1020 10:10:06.502223 11597 processor.cc:42] processor_tid: 11597
I1020 10:10:06.503582 11598 processor.cc:42] processor_tid: 11598
I1020 10:10:06.504865 11599 processor.cc:42] processor_tid: 11599
I1020 10:10:06.505489 11600 processor.cc:42] processor_tid: 11600
I1020 10:10:06.506629 11601 processor.cc:42] processor_tid: 11601
I1020 10:10:06.507230 11602 processor.cc:42] processor_tid: 11602
I1020 10:10:06.508342 11603 processor.cc:42] processor_tid: 11603
I1020 10:10:06.509541 11604 processor.cc:42] processor_tid: 11604
I1020 10:10:06.510165 11605 processor.cc:42] processor_tid: 11605
I1020 10:10:06.511026 11606 processor.cc:42] processor_tid: 11606
I1020 10:10:06.511652 11607 processor.cc:42] processor_tid: 11607
I1020 10:10:06.512698 11608 processor.cc:42] processor_tid: 11608
I1020 10:10:06.513756 11609 processor.cc:42] processor_tid: 11609
I1020 10:10:06.516114 11584 init.cc:113] Register exit handle succ.
I1020 10:10:06.522330 11584 participant.cc:121] [DEBUG] cyber ip: 127.0.0.1
I1020 10:10:06.603684 11584 notifier_factory.cc:40] [DEBUG] notifier type: condition
I1020 10:10:06.603962 11584 condition_notifier.cc:34] [DEBUG] condition notifier key: 360705043
I1020 10:10:06.604301 11584 condition_notifier.cc:156] [DEBUG] open or create true.
I1020 10:10:06.604413 11584 condition_notifier.cc:43] [DEBUG] next_seq: 0
I1020 10:10:06.605490 11584 notifier_factory.cc:40] [DEBUG] notifier type: condition
I1020 10:10:06.626740 11584 rtps_transmitter.h:102] [DEBUG] not enable.
I1020 10:10:06.626878 11584 intra_transmitter.h:80] [DEBUG] not enable.
I1020 10:10:06.626994 11584 shm_transmitter.h:103] [DEBUG] not enable.
I1020 10:10:06.627089 11584 talker.cc:43] talker sent a message! No. 0
I1020 10:10:06.825300 11616 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:07.059653 11616 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:07.289197 11616 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:07.517704 11616 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:07.627179 11584 rtps_transmitter.h:102] [DEBUG] not enable.
I1020 10:10:07.627336 11584 intra_transmitter.h:80] [DEBUG] not enable.
I1020 10:10:07.627440 11584 shm_transmitter.h:103] [DEBUG] not enable.
I1020 10:10:07.627532 11584 talker.cc:43] talker sent a message! No. 1
I1020 10:10:07.746855 11616 shm_dispatcher.cc:106] [DEBUG] listen failed.

listener log
Log file created at: 2020/10/20 10:10:13
Running on machine: in-cyber-docker
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
I1020 10:10:13.718183 11660 global_data.cc:153] host ip: 192.168.10.31
W1020 10:10:13.773814 11660 scheduler_factory.cc:63] No sched conf found, use default conf.
I1020 10:10:13.782703 11664 processor.cc:42] processor_tid: 11664
I1020 10:10:13.785434 11665 processor.cc:42] processor_tid: 11665
I1020 10:10:13.787539 11666 processor.cc:42] processor_tid: 11666
I1020 10:10:13.788728 11667 processor.cc:42] processor_tid: 11667
I1020 10:10:13.789942 11668 processor.cc:42] processor_tid: 11668
I1020 10:10:13.791162 11669 processor.cc:42] processor_tid: 11669
I1020 10:10:13.792448 11670 processor.cc:42] processor_tid: 11670
I1020 10:10:13.793028 11671 processor.cc:42] processor_tid: 11671
I1020 10:10:13.793951 11672 processor.cc:42] processor_tid: 11672
I1020 10:10:13.794981 11673 processor.cc:42] processor_tid: 11673
I1020 10:10:13.795971 11674 processor.cc:42] processor_tid: 11674
I1020 10:10:13.796630 11675 processor.cc:42] processor_tid: 11675
I1020 10:10:13.797479 11676 processor.cc:42] processor_tid: 11676
I1020 10:10:13.798988 11677 processor.cc:42] processor_tid: 11677
I1020 10:10:13.799687 11678 processor.cc:42] processor_tid: 11678
I1020 10:10:13.801323 11679 processor.cc:42] processor_tid: 11679
I1020 10:10:13.803562 11660 init.cc:113] Register exit handle succ.
I1020 10:10:13.809839 11660 participant.cc:121] [DEBUG] cyber ip: 127.0.0.1
I1020 10:10:13.900394 11660 scheduler.cc:55] create croutine: listener_channel/chatter
I1020 10:10:13.905295 11660 notifier_factory.cc:40] [DEBUG] notifier type: condition
I1020 10:10:13.905547 11660 condition_notifier.cc:34] [DEBUG] condition notifier key: 360705043
I1020 10:10:13.905694 11660 condition_notifier.cc:126] [DEBUG] shm already exist, open only.
I1020 10:10:13.906064 11660 condition_notifier.cc:184] [DEBUG] open true.
I1020 10:10:13.906395 11660 condition_notifier.cc:43] [DEBUG] next_seq: 0
I1020 10:10:13.908100 11660 notifier_factory.cc:40] [DEBUG] notifier type: condition
I1020 10:10:14.128000 11686 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:14.362041 11686 shm_dispatcher.cc:106] [DEBUG] listen failed.
I1020 10:10:14.591575 11686 shm_dispatcher.cc:106] [DEBUG] listen failed.

Do you know the reason why it didn't work properly?

I ran this test on x86 machine.
Host machine x86 and target machine aarch64.

aarch64 container was working on top of x86 host machine.

**Comment by storypku** [comment]
I personally have no clue at present. Maybe it's related to cross-architecture shared memory issue of Docker when `cyber_start.sh`?

Maybe you can tweak `cyber_start.sh` by designating   `--ipc=host` for docker run to take place of `--shm-size` to discover more.

**Comment by storypku** [comment]
This issue is going to be closed since cross-platform support for Apollo (running Apollo aarch64 docker image on top of x86 host, or vise versa),  will not be supported by the Apollo team. However, PRs from the community are always welcome.

If you have other questions, please feel free to open another issue. Or reopen this one if you feel the need to discuss more.
